### ISSUE TITLE
Recipe for updating DOM structure based on Receiver

### ISSUE THREAD

**Issue by bryanjswift** [schell/mogwai#43]
While investigating how I would implement something with multiple pages (refs #32) I was trying to set up an enum of routes and have different a different structure/view depending on the route. Looking at the `ViewBuilder` it looks like `Effect` and `Receiver` can only be used to create a `ViewBuilder` if the type argument is a string value (so a text node) and I could not find an example where something other than text was being updated based on `txrx` functions. For reference, below is (an iteration) of what I was trying (based on 0081a29); I'm fairly sure this is incorrect conceptually because it isn't seeing an updated value in the `Receiver` so I'm not trying to fix the code below but rather figure out how it should have been written to work with `Receiver` correctly.

```rust
use crate::routes;
use log::trace;
use mogwai::prelude::*;

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Route {
    Home,
    NotFound,
}

#[derive(Debug)]
pub struct App {
    current_route: Route,
}

impl App {
    pub fn new() -> Self {
        App {
            current_route: Route::Home,
        }
    }
}

impl Component for App {
    type ModelMsg = Route;
    type ViewMsg = ();
    type DomNode = HtmlElement;

    fn update(&mut self, msg: &Route, tx_view: &Transmitter<()>, _sub: &Subscriber<Route>) {
        trace!("{:?}", msg);
        self.current_route = msg.clone();
        tx_view.send(&());
    }

    fn view(&self, tx: &Transmitter<Route>, _rx: &Receiver<()>) -> ViewBuilder<HtmlElement> {
        trace!("{:?}", &self);
        let main = match self.current_route {
            // pub fn home() -> ViewBuilder<HtmlElement>
            Route::Home => routes::home(),
            // pub fn not_found() -> ViewBuilder<HtmlElement>
            Route::NotFound => routes::not_found(),
        };
        builder! {
            <div class="root">
                <nav>
                    <button on:click=tx.contra_map(|_| Route::Home)>
                        "Home"
                    </button>
                    <button on:click=tx.contra_map(|_| Route::NotFound)>
                        "Not Found"
                    </button>
                </nav>
                <main>
                    {main}
                </main>
            </div>
        }
    }
}
```

**Comment by bryanjswift** [comment]
I think it _should_ be something like this

```rust
use crate::routes;
use log::trace;
use mogwai::prelude::*;

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Route {
    Home,
    NotFound,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Out {
    Render(Route),
    SkipRender,
}

#[derive(Debug)]
pub struct App {
    current_route: Route,
}

impl App {
    pub fn new() -> Self {
        App {
            current_route: Route::Home,
        }
    }

    fn render(route: Route) -> ViewBuilder<HtmlElement> {
        match route {
            // pub fn home() -> ViewBuilder<HtmlElement>
            Route::Home => routes::home(),
            // pub fn not_found() -> ViewBuilder<HtmlElement>
            Route::NotFound => routes::not_found(),
        }
    }
}

impl Component for App {
    type ModelMsg = Route;
    type ViewMsg = Out;
    type DomNode = HtmlElement;

    fn update(&mut self, msg: &Route, tx_view: &Transmitter<Out>, _sub: &Subscriber<Route>) {
        trace!("{:?}", msg);
        self.current_route = msg.clone();
        if self.current_route == *msg {
            tx_view.send(&Out::SkipRender);
        } else {
            tx_view.send(&Out::Render(*msg));
        }
    }

    fn view(&self, tx: &Transmitter<Route>, rx: &Receiver<Out>) -> ViewBuilder<HtmlElement> {
        trace!("{:?}", &self);
        let rx_main = rx
            .branch_fold(self.current_route, |current_route, msg| match msg {
                Out::Render(route) => *route,
                _ => *current_route,
            })
            .branch_map(|route| App::render(*route));
        builder! {
            <div class="root">
                <nav>
                    <button on:click=tx.contra_map(|_| Route::Home)>
                        "Home"
                    </button>
                    <button on:click=tx.contra_map(|_| Route::NotFound)>
                        "Not Found"
                    </button>
                </nav>
                <main>
                    {rx_main}
                </main>
            </div>
        }
    }
}
```

but this doesn't compile because 

```
the trait `std::convert::From<mogwai::prelude::Receiver<mogwai::prelude::ViewBuilder<web_sys::HtmlElement>>>` is not implemented for `mogwai::prelude::ViewBuilder<_>`
```

**Comment by bryanjswift** [comment]
The only thing close I've seen is the TodoMVC implementation that appends child nodes to the `<ul>` during `post:build` but having to add a `post:build` to the `<main>` element here and append a `View` does not seem like the desired experience.

**Comment by schell** [comment]
I agree with you that this is not the ideal experience. That is, I do not like having to patch the DOM from within the update function explicitly, because it hides information about what the DOM will be in the future.

I have two new interfaces (traits in src/view/interface.rs) Called ReplaceView and PatchView That allow a DOM node to be specified by a receiver and allow updates to the children of a DOM node to be specified by a receiver, respectively. 

These two interfaces are still being worked on in a separate branch but I plan to make them part of the next release. I think they will serve your purpose directly.

Thanks for hanging in there while I figure out what the best API for Mogwai is!

**Comment by bryanjswift** [comment]
That’s great! I saw those traits in one of the recent commits in #40 but did not grok their usage. After the series of comments here I had another try with a `ViewBuilder` with element and replacers defined which seemed like it was closer to the outcome I wanted if a bit more verbose. I’m glad to hear this use case is on the radar and to be addressed in the next release! I’ll be on the lookout for those changes to land!

**Comment by schell** [comment]
Once those changes land (and you can preview them in the `less_allocy_builder` branch) I think your `impl Component for App` would look something like:  
```rust 

impl Component for App {
    type ModelMsg = Route;
    type ViewMsg = Out;
    type DomNode = HtmlElement;

    fn update(&mut self, msg: &Route, tx_view: &Transmitter<Out>, _sub: &Subscriber<Route>) {
        trace!("{:?}", msg);
        self.current_route = msg.clone();
        if self.current_route == *msg {
            tx_view.send(&Out::SkipRender);
        } else {
            tx_view.send(&Out::Render(*msg));
        }
    }

    fn view(&self, tx: &Transmitter<Route>, rx: &Receiver<Out>) -> ViewBuilder<HtmlElement> {
        trace!("{:?}", &self);
        let rx_main = rx
            .branch_fold(self.current_route, |current_route, msg| match msg {
                Out::Render(route) => *route,
                _ => *current_route,
            })
            .branch_map(|route| View::from(App::render(*route)));
        builder! {
            <div class="root">
                <nav>
                    <button on:click=tx.contra_map(|_| Route::Home)>
                        "Home"
                    </button>
                    <button on:click=tx.contra_map(|_| Route::NotFound)>
                        "Not Found"
                    </button>
                </nav>
                <main this:later=rx_main>
                    "Placeholder content"
                </main>
            </div>
        }
    }
}
```

**Comment by schell** [comment]
Just to be 100% explicit, the RSX `<main this:later=rx_main>"..."</main>` means "start with `<main>...</main>` and then replace the outer html with whatever is sent on the receiver `rx_main` - so the entire `<main>` node is replaced with the message of the receiver.

**Comment by bryanjswift** [comment]
> Just to be 100% explicit, the RSX `<main this:later=rx_main>"..."</main>` means "start with `<main>...</main>` and then replace the outer html with whatever is sent on the receiver `rx_main` - so the entire `<main>` node is replaced with the message of the receiver.

Right, that was my read of this as well. So each "route" needs to render a `View` containing the `<main>...</main>` element if that is intended to be a "global" wrapper.

**Comment by bryanjswift** [comment]
Is the intention that `rx_main` (to continue this example) will need to be a `Receiver<View<T>>` or could it also be a `Receiver` of anything that can be turned into a `View`?

**Comment by schell** [comment]
Right now it's a `Receiver<View<T>>` but maybe I should play around with it because that interface may be easier to work with. It depends on how easy it is for the type system to figure out the `T`.

**Comment by schell** [comment]
It ended up not being difficult to provide that API https://github.com/schell/mogwai/pull/40/commits/f9db02ab0d386c429a42b479db92cee359e6ff37. Hopefully it works at the callsites - it seems to from my testing.

**Comment by bryanjswift** [comment]
I'd call this closed with the merge of #40 and the additions to examples and cookbook.
